configserver/src/main/java/com/yahoo/vespa/config/server/maintenance/ApplicationPackageMaintainer.java

```diff
@@ -64,49 +64,55 @@ protected double maintain() {
+                attempts++;
+                if (!fileReferenceExistsOnDisk(downloadDirectory, appFileReference)) {
+                    log.fine(() -> "Downloading application package for " + applicationId + " (session " + sessionId + ")");
+
+                    FileReferenceDownload download = new FileReferenceDownload(appFileReference,
+                                                                               false,
+                                                                               this.getClass().getSimpleName());
+                    if (fileDownloader.getFile(download).isEmpty()) {
+                        failures++;
+                        log.info("Failed downloading application package (" + appFileReference + ")" +
+                                         " for " + applicationId + " (session " + sessionId + ")");
+                        continue;
                     }
-                    createLocalSessionIfMissing(applicationId, sessionId);
                 }
+                createLocalSessionIfMissing(applicationId, sessionId);
             }
         }
         return asSuccessFactor(attempts, failures);
     }
 
-    private FileDownloader createFileDownloader() {
-        ConfigSourceSet configSourceSet = new ConfigSourceSet(getOtherConfigServersInCluster(configserverConfig));
-        return new FileDownloader(useFileDistributionConnectionPool
-                                          ? new FileDistributionConnectionPool(configSourceSet, supervisor)
-                                          : new JRTConnectionPool(configSourceSet, supervisor),
-                                  supervisor,
-                                  downloadDirectory,
-                                  Duration.ofSeconds(30));
+    private static FileDownloader createFileDownloader(ConfigserverConfig configserverConfig,
+                                                       boolean useFileDistributionConnectionPool,
+                                                       File downloadDirectory) {
+        Supervisor supervisor = new Supervisor(new Transport("filedistribution-pool")).setDropEmptyBuffers(true);
+        List<String> otherConfigServersInCluster = getOtherConfigServersInCluster(configserverConfig);
+        ConfigSourceSet configSourceSet = new ConfigSourceSet(otherConfigServersInCluster);
+
+        ConnectionPool connectionPool;
+        if (otherConfigServersInCluster.isEmpty())
+            connectionPool = FileDownloader.emptyConnectionPool();
+        else
+            connectionPool = useFileDistributionConnectionPool
+                    ? new FileDistributionConnectionPool(configSourceSet, supervisor)
+                    : new JRTConnectionPool(configSourceSet, supervisor);
+        return new FileDownloader(connectionPool, supervisor, downloadDirectory, Duration.ofSeconds(30));
     }
 
     @Override
     public void awaitShutdown() {
-        supervisor.transport().shutdown().join();
```

Comment:
Don't we have to shutdown the transport thread now?

Reply:
It's done by `FileDownloader.close()`, but I agree this does not look good. Putting it back again, I'll look into other users of FileDownloader and see if this can be done better in another PR

Reply:
Thanks, it looks much better now.